### PR TITLE
Wrap custom iterator result

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
Fixes: #17165 
Fixes: https://github.com/rapidsai/cudf/issues/14481

This PR properly wraps the result of custom iterator. 

```python
In [2]: import pandas as pd

In [3]: s = pd.Series([10, 1, 2, 3, 4, 5]*1000000)


# Without custom_iter:

In [4]: %timeit for i in s: True
6.34 s ± 25.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# This PR:

In [4]: %timeit for i in s: True
6.16 s ± 17.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


# On `branch-24.12`:
1.53 s ± 6.27 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I think `custom_iter` has to exist. Here is why, invoking any sort of `iteration` on GPU objects will raise errors and thus in the end we fall-back to CPU. Instead of trying to move the objects from host to device memory (if the object is on host memory only), we will avoid a CPU-to-GPU transfer.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by galipremsagar** [comment]
> @galipremsagar could you add a test that shows the undesirable fallback behavior of iter? You should be able to do this by monkey-patching `os.environ` with [`CUDF_PANDAS_FAIL_ON_FALLBACK`](https://github.com/rapidsai/cudf/blob/branch-24.12/python/cudf/cudf/pandas/fast_slow_proxy.py#L966) and showing that fallback is triggered undesirably. That ought to be possible without a subprocess, but you might have to define a new custom proxy type without `custom_iter` to show it.
> 
> If that turns out to be too much work we can just document it, but I want to make it as obvious as possible to future devs why this is necessary.

I agree, I'll add a test and document the reason behind the custom iterator existence.

**Comment by vyasr** [comment]
This PR also closes #14481 right?

**Comment by galipremsagar** [comment]
> I agree, I'll add a test and document the reason behind the custom iterator existence.


Done. 



> This PR also closes #14481 right?

Yes, this also closes #14481

**Comment by galipremsagar** [comment]
@vyasr @Matt711 This is now ready for review.

File: python/cudf/cudf/pandas/fast_slow_proxy.py [hunk 1]
```diff
@@ -1099,7 +1099,9 @@ def _maybe_wrap_result(result: Any, func: Callable, /, *args, **kwargs) -> Any:
     """
     Wraps "result" in a fast-slow proxy if is a "proxiable" object.
     """
-    if _is_final_type(result):
+    if isinstance(result, (int, str, float, bool, type(None))):
+        return result
+    elif _is_final_type(result):
```
Comment (Matt711) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1104]: Should we incorporate this check in `_is_final_type`?

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py [hunk 2]
```diff
@@ -1777,3 +1777,22 @@ def test_cudf_pandas_util_version(attrs):
...
         assert hasattr(pd.util, attrs)
+
+
+def test_iteration_over_dataframe_dtypes_produces_proxy_objects(dataframe):
+    _, xdf = dataframe
+    xdf["b"] = xpd.IntervalIndex.from_arrays(xdf["a"], xdf["b"])
+    xdf["a"] = xpd.Series([1, 1, 1, 2, 3], dtype="category")
+    dtype_series = xdf.dtypes
+    for x in dtype_series:
+        assert is_proxy_object(x)
```
Comment (Matt711) [inline python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1788]:
```suggestion
    assert all(is_proxy_object(x) for x in dtype_series)
```

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py [hunk 3]
```diff
@@ -1777,3 +1777,22 @@ def test_cudf_pandas_util_version(attrs):
...
+    assert isinstance(dtype_series.iloc[0], xpd.CategoricalDtype)
+    assert isinstance(dtype_series.iloc[1], xpd.IntervalDtype)
+
+
+def test_iter_doesnot_raise(monkeypatch):
+    s = xpd.Series([1, 2, 3])
+    with monkeypatch.context() as monkeycontext:
+        monkeycontext.setenv("CUDF_PANDAS_FAIL_ON_FALLBACK", "True")
+        for i in s:
```
Comment (Matt711) [inline python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1797]:
```suggestion
        for _ in s:
```

**Review by Matt711 — COMMENTED** [review]
A question and tiny suggestions

File: python/cudf/cudf/pandas/fast_slow_proxy.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1104]: I don't think it is a good idea to add it to `_is_final_type`, those are for pandas types and the ones here we are trying to capture are native python types.

**Review by Matt711 — APPROVED** [review]

**Comment by galipremsagar** [comment]
Going ahead and merging this PR. @vyasr incase you have additional comments let me know I can address in a follow-up.

**Comment by vyasr** [comment]
No worries, this looks great thanks!

**Pull request merged**